src/compute-client/src/controller/replicated.rs

```diff
@@ -188,34 +189,45 @@ where
         comm_config,
     };
 
-    loop {
-        select! {
-            // Command from controller to forward to replica.
-            command = command_rx.recv() => match command {
-                None => {
-                    // Controller is no longer interested in this replica. Shut
-                    // down.
-                    return Ok(())
-                }
-                Some(mut command) => {
-                    cmd_spec.specialize_command(&mut command);
-                    client.send(command).await?
-                }
-            },
-            // Response from replica to forward to controller.
-            response = client.recv() => {
-                let response = match response? {
-                    None => bail!("replica unexpectedly gracefully terminated connection"),
-                    Some(response) => response,
-                };
-                if response_tx.send(response).is_err() {
-                    // Controller is no longer interested in this replica. Shut
-                    // down.
-                    return Ok(());
+    let res = async {
+        loop {
+            select! {
+                // Command from controller to forward to replica.
+                command = command_rx.recv() => match command {
+                    None => {
+                        // Controller is no longer interested in this replica. Shut down.
+                        return Ok(());
+                    }
+                    Some(mut command) => {
+                        if let ComputeCommand::DropInstance = command {
+                            // Controller is no longer interested in this replica. Shut down.
+                            return Ok(());
+                        };
+                        cmd_spec.specialize_command(&mut command);
+                        client.send(command).await?
+                    }
+                },
+                // Response from replica to forward to controller.
+                response = client.recv() => {
+                    let response = match response? {
+                        None => bail!("replica unexpectedly gracefully terminated connection"),
+                        Some(response) => response,
+                    };
+                    if response_tx.send(response).is_err() {
+                        // Controller is no longer interested in this replica. Shut down.
+                        return Ok(());
+                    }
                 }
             }
         }
     }
+    .await;
+
+    if let ComputeReplicaLocation::Managed { .. } = location {
+        orchestrator.drop_replica(instance_id, replica_id).await?;
+    }
```

Comment:
Agree, I changed this now, the BG task will only remove from the orchestrator if it receives a DropInstance, otherwise it assumes that the reconciliation logic will restart the task.